server/session_ws.go

```diff
@@ -173,7 +175,7 @@ func (s *sessionWS) Consume(processRequest func(logger *zap.Logger, session Sess
 		s.receivedMessageCounter--
 		if s.receivedMessageCounter <= 0 {
 			s.receivedMessageCounter = s.config.GetSocket().PingBackoffThreshold
-			s.pingTimer.Reset(s.pingPeriodDuration)
+			s.pingTimerResetCh <- s.pingPeriodDuration
```

Comment:
The ping backoff exists to avoid unnecessary context switches caused by session ping/pong at high concurrent user counts when the socket is known to be constantly active. Introducing a channel here might bring back this impact, but I definitely see the case where concurrent access is possible and should be prevented.

What do you think of using https://github.com/heroiclabs/nakama/commit/ac9810151dccc9ef4db8e2b2d893c08158996caf to achieve this instead?

Reply:
I see your point, I believe https://github.com/heroiclabs/nakama/commit/7a3a800d2fcfb34360712189e3307bc09358b159 handles it correctly then. Thoughts?

Reply:
Awesome, it's correct now. I've just read [time.go](https://golang.org/src/runtime/time.go) source code and the problematic session_ws.go. The root cause of the panics was timer.Reset is called in both receiving goroutine and sending goroutine and will cause a racy access [as this gist shown](https://gist.github.com/zengjie/ec6194c5f8d2b6eefc2e2a8340c29e7f). Now, we have only resets in receiving goroutine, there will be no panic.

However, I am wondering how to maintain ping period if we don't reset the timer in pingNow. I think it's ok to reset in pingNow with the proposed "stop and check" protection.

Reply:
The timer will be reset when the corresponding pong is received. If the pong never makes it then we're closing the session anyway and there's no point in resetting the timer.

I think we have our solution then!